### PR TITLE
`azurerm_container_registry` - support update replications on demand

### LINKED ISSUE #16634: Geo-replication regions deleted/recreated when modified
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.0.11

### AzureRM Provider Version

2.99

### Affected Resource(s)/Data Source(s)

azurerm_container_registry

### Terraform Configuration Files

```hcl
resource "azurerm_container_registry" "acr" {
  # Whatever other settings, but has to be a premium registry
  georeplication_locations      = [
    "westus",
    "canadaeast",
    "canadacentral",
    "uksouth",
    "ukwest",
    "northeurope",
    "westeurope",
    "germanywestcentral",
    "australiaeast",
    "australiasoutheast"
  ]
}
```

### Debug Output/Panic Output

```shell
A plan after removing 'australiasoutheast' produces the following output:

# azurerm_container_registry.acr will be updated in-place
  ~ resource "azurerm_container_registry" "acr" {
      ~ georeplication_locations      = [
          - "australiasoutheast",
            # (9 unchanged elements hidden)
        ]

An Apply will remove/recreate all the geo-replications causing temporary downtime, in our case it was even worse because the apply failed halfway through (deleting some geo-replications we did not intend on touching at all)
```

### Expected Behaviour

When 'australiasoutheast' (or any other region) is removed from the list of Geo-replications, it should only delete the targeted replication, or at least indicate that ALL replications will be deleted/recreated (see https://github.com/hashicorp/terraform-provider-azurerm/blob/35a2aa81e342d142b5152f83511e8d36278099b7/internal/services/containers/container_registry_resource.go#L434)

### Actual Behaviour

When deleting one or more of the geo-replications, it deletes/recreates all of them, potentially causing downtime and other issues for big container registries. In our case the apply failed halfway through, causing unwanted deletion of some Geo-replications that we did not intend to touch.

### Steps to Reproduce

1. Terraform plan/apply with the above Geo-replication regions
2. Terraform plan/apply with one region commented out

### Important Factoids

_No response_

### References

### PR DESCRIPTION
**Pull request by magodo**
This PR adds more finer grained control on how to update the replications for `azurerm_container_registry`.

Previously, any change in the `georeplications` will cause all replications to be recreated. This causes unexpected downtime for replications that are not affected (see #16634). This PR fixes it in way that:

- Only delete replications that are deleted from the config
- Only create replications that are newly added into the config
- Only Update replications that are updated in the config
- Otherwise, do nothing to the replication

Specially for update replications, API currently doesn't support updating the `zoneRedundancy` in place, which makes us have to delete then create it. Also, API has a bug when create a just deleted replication, see: https://github.com/Azure/azure-rest-api-specs/issues/18934. Therefore, I have added a explicit wait in this PR to avoid that.

Additionally, during the test I noticed the LIST api of the replication will return the replications in arbitrary order. I have sorted it for the `georeplications` alphabetically during flattening. So that make the plan diff more readable for users (assuming users order their config alphabetically as well). Ideally, it should use `Set` instead, but [it can't](https://github.com/hashicorp/terraform-provider-azurerm/compare/main...magodo:acr_geo_update?expand=1#diff-bf653d131a0b6548f81b0a4282f73e691d7e776f6147c88a49c15977d562eab6R1048).

Fixes #16634.

## Test

```shell
💢  TF_ACC=1 go test -v -timeout=20h ./internal/services/containers -run='TestAccContainerRegistry_geoReplication|TestAccContainerRegistry_geoReplicationLocation'
=== RUN   TestAccContainerRegistry_geoReplicationLocation
=== PAUSE TestAccContainerRegistry_geoReplicationLocation
=== RUN   TestAccContainerRegistry_geoReplication
=== PAUSE TestAccContainerRegistry_geoReplication
=== RUN   TestAccContainerRegistry_geoReplicationZoneRedundancy
=== PAUSE TestAccContainerRegistry_geoReplicationZoneRedundancy
=== RUN   TestAccContainerRegistry_geoReplicationRegionEndpoint
=== PAUSE TestAccContainerRegistry_geoReplicationRegionEndpoint
=== CONT  TestAccContainerRegistry_geoReplicationLocation
=== CONT  TestAccContainerRegistry_geoReplicationZoneRedundancy
=== CONT  TestAccContainerRegistry_geoReplicationRegionEndpoint
=== CONT  TestAccContainerRegistry_geoReplication
--- PASS: TestAccContainerRegistry_geoReplicationRegionEndpoint (259.38s)
--- PASS: TestAccContainerRegistry_geoReplicationZoneRedundancy (350.45s)
--- PASS: TestAccContainerRegistry_geoReplicationLocation (1235.67s)
--- PASS: TestAccContainerRegistry_geoReplication (1965.56s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/containers    1965.580s
```

Additionally, I've locally inspected the API sequence and confirmed it worked as expected.

### CODE REVIEW

File: website/docs/r/container_registry.html.markdown [hunk 1]
```diff
@@ -188,6 +188,8 @@ The following arguments are supported:
 
 * `zone_redundancy_enabled` - (Optional) Whether zone redundancy is enabled for this replication location? Defaults to `false`.
 
+  ~> **NOTE:** Changing the `zone_redundancy_enabled` forces the a new underlying replication to be created.
```
Comment (katbyte) [inline website/docs/r/container_registry.html.markdown:191]:
```suggestion
  ~> **NOTE:** Changing the `zone_redundancy_enabled` forces the a underlying replication to be created.
```

**Review by katbyte — APPROVED** [review]
Thanks @magodo - LGTM 🐍

**Comment by katbyte** [comment]
@magodo - could we fix up the merge conflics to get this merged? thanks

**Comment by magodo** [comment]
@katbyte Sure, I've resolved it. Thank you for the review!

**Pull request merged**